### ISSUE TITLE
Make enumeratum part of the SP (Scala Platform)

### ISSUE THREAD

**Issue by mdedetrich** [lloydmeta/enumeratum#92]
With the new Scala Platform being announced (see https://contributors.scala-lang.org/) we now have the ability to add libraries to a defined Scala platform.

Enumeratum fills a pretty core functionality (creating proper Enum's out of ADT's) which would conceivably be required by a lot of libraries (including one that I am writing right now). Its also probably the most well tested and used library in this specific area.

The only possible issues that I can envisage is what parts of enumeratum would be part of the Scala Platform. I envisage that only enumeratum core would be included (because there is a pretty strict requirement regarding transitive dependencies). If this was to go ahead (and its decided that only enumeratum-core would go into SP), we would probably have to

- Add in MIMA to enumeratum-core to ensure binary compatibility
- Change enumeratum-core's versioning scheme to be different from the various addons, as the enumeratum-core versioning scheme should reflect its own stability (i.e. enumeratum-core's version number shouldn't be bumped because a new version of enumeratum-circe is released which has no bearing on core)
- May need to rename enumeratum-core to something else if we have versioning conflicts due to changing the versioning scheme
- Possibly make enumartum-core its own actual project? Not sure if this is required.
- All the various consequences of being part of the Scala platform (described on the link earlier)

What are your thoughts on this? I am also willing to do the heavy lifting if you generally agree with the proposal in general.

**Comment by lloydmeta** [comment]
I think this is a good idea, thanks for bringing it up. In general, the suggestions you bring up make sense anyway, such as adding MIMA and decoupling the versions between the different libraries, esp. core (to be honest, this was more of a convenience thing for me).

I took a look at the [Scala Platform policies](https://scalacenter.github.io/platform-staging/policies.html) and am a bit concerned that there seem to be a lot of rules and procedures that libraries that are part of the platform need to adhere to. They all make sense (I like stability a lot too), but it's a bit much for me to do all by myself. If Enumeratum joins the SP, I'll at least need another maintainer to help, or even take over the reins for a bit because I'm afraid I might make mistakes in the beginning.

So in short, I agree with the idea, as long as you (or someone else ?) can help out after the transition as well. I will of course take up your offer of doing the heavy lifting :)

Actually the main question I have is: what do *I* need to do to help make this happen?

**Comment by mdedetrich** [comment]
> I took a look at the Scala Platform policies and am a bit concerned that there seem to be a lot of rules and procedures that libraries that are part of the platform need to adhere to.

I believe these are open to debate (you can actually talk about them on the forum I linked earlier). Some of the rules (such as license) aren't actually rules, but mere recommendations. I think the strict rules are about things like stability (must support latest 2 stable versions of scala) and having your project be built by the scala build project

> Actually the main question I have is: what do I need to do to help make this happen?

I think for now its mainly the technical things that were mentioned earlier (MIMA, seperate version for enumeratum-core) and making a thread on scala contributors indicating that you want it to be part of the Scala Platform. Adding the project to be built part of the Scala Build Platform also isn't a lot of work

**Comment by lloydmeta** [comment]
> I believe these are open to debate (you can actually talk about them on the forum I linked earlier). Some of the rules (such as license) aren't actually rules, but mere recommendations.

Ah ok. In that case, I think we can move ahead and see where the discussion on Scala Contributors takes us :+1:

**Comment by jvican** [comment]
Thanks @mdedetrich and @lloydmeta for considering the inclusion of enumeratum in the SP, that is a great idea!
  
> Actually the main question I have is: what do I need to do to help make this happen?
  
I'll reply to this in the [Scala Contributors thread you've opened](https://contributors.scala-lang.org/t/adding-enumeratum-to-the-scala-platform/307). For the moment, you don't need to change anything in your build :wink:. Please, do comment on any of the open threads discussing the Scala Platform, your opinion is key.
  
> I think for now its mainly the technical things that were mentioned earlier (MIMA, seperate version for enumeratum-core)
  
`sbt-platform` is meant to alleviate the effort of including MiMa in your build and allow automatic releases (details [here](https://scalacenter.github.io/platform-staging/sbt-platform.html)). If we incubate `enumeratum`, I will integrate `sbt-platform` into your build and make sure it works, so hopefully you won't need to do anything.

**Comment by mdedetrich** [comment]
I think the first step that we need to do is to split out the enumeratum-core (especially in regards to versioning) since we don't want to bump the version of enumeratum-core every time an addon is released. After that point we can include it into sbt-platform

**Comment by jvican** [comment]
Indeed, that would be necessary @mdedetrich. BTW, is `enumeratum` using semantic versioning?

**Comment by mdedetrich** [comment]
> BTW, is enumeratum using semantic versioning?

No idea, but I think that @lloydmeta should clarify this (we may need to change the versioning scheme as well then)

**Comment by lloydmeta** [comment]
The versioning is *sorta* semantic, because as @mdedetrich noted before, I've been bumping the version of every library (incl. core) whenever a new version of the whole suite in lockstep (for the sake of convenience at release). I think this lock-step bumping is something as we've all agreed something that should be changed.

I think it may be easiest to start by having each project use its own version, without splitting enumeratum-core out into a separate project (source control / repository-wise). After all, the projects are already all in their own self-defined sub-projects, which means they already have their own `:version` setting key.

**Comment by akauppi** [comment]
How is this connected with the Enumeration rise-again of Scala 2.13? https://www.scala-lang.org/api/current/scala/Enumeration.html

**Comment by lloydmeta** [comment]
We're very likely not going to be making Enumeratum part of SP since the proposal thread didn't gain traction.
